csharp-mode.el

```diff
@@ -2077,1076 +1837,282 @@ to the beginning of the prior namespace.
+                        "\\)"
+                        optional-space
+                        ";") 1))))
+
+(defun csharp--imenu-get-pos (pair)
+  (let ((pos (cdr pair)))
+    (if (markerp pos)
+        (marker-position pos)
+      pos)))
+
+(defun csharp--imenu-get-container (item containers previous)
+  (if (not containers)
+      previous
+    (let* ((item-pos (csharp--imenu-get-pos item))
+           (container (car containers))
+           (container-pos (csharp--imenu-get-pos container))
+           (rest      (cdr containers)))
+      (if (and container-pos
+               (< item-pos container-pos))
+          previous
+        (csharp--imenu-get-container item rest container)))))
+
+(defun csharp--imenu-get-container-name (item containers)
+  (let ((container (csharp--imenu-get-container item containers nil)))
+    (if (not container)
+        nil
+      (let* (;; namespace
+             (container-p1 (car (split-string (car container))))
+             ;; class/interface
+             (container-p2 (cadr (split-string (car container)))))
+        ;; use p1 (namespace) when there is no p2
+        (if container-p2
+            container-p2
+          container-p1)))))
+
+(defun csharp--imenu-sort (items)
+  (sort items (lambda (item1 item2)
+                (string< (car item1) (car item2)))))
+
+(defun csharp--imenu-get-class-name (class namespaces)
+  (let ((namespace (csharp--imenu-get-container-name class namespaces))
+        (class-name (car class)))
+    (if (not namespace)
+        class-name
+      ;; reformat to include namespace
+      (let* ((words (split-string class-name))
+             (type  (car words))
+             (name  (cadr words)))
+        (concat type " " namespace "." name)))))
+
+(defun csharp--imenu-get-class-nodes (classes namespaces)
+  "Creates a new alist with classes as root nodes with namespaces added."
+
+  (mapcar (lambda (class)
+            (let* ((class-name (csharp--imenu-get-class-name class namespaces))
+                   (class-pos  (cdr class)))
+              (cons class-name
+                    (list
+                     (cons "( top )" class-pos)))))
+          classes))
+
+(defun csharp--imenu-get-class-node (result item classes namespaces)
+  (let* ((class-item (csharp--imenu-get-container item classes nil))
+         (class-name (csharp--imenu-get-class-name class-item namespaces)))
+    (assoc class-name result)))
+
+(defun csharp--imenu-format-item-node (item type)
+  (cons
+   (concat "(" type ") " (car item))
+   (cdr item)))
+
+(defun csharp--imenu-append-items-to-menu (result type index classes namespaces)
+  ;; items = all methods, all events, etc based on "type"
+  (let* ((items (cdr (assoc type index))))
+    (dolist (item items)
+      (let ((class-node (csharp--imenu-get-class-node result item classes namespaces))
+            (item-node  (csharp--imenu-format-item-node item type)))
+        (nconc class-node (list item-node))))))
```

Comment:
I'm looking to *destructively* mutate the list I've been handed as a parameter so that the holder of the original list gets the modifications.

For that nconc seems like the best option, as far as I can see. Or am I missing something?